crates/bitwarden-crypto/src/signing/sign.rs

```diff
@@ -0,0 +1,588 @@
+//!   Purpose: All signatures should be domain-separated, so that any proofs only need to consider
+//!   the allowed messages under the current namespace, and cross-protocol attacks are not possible.
+//! - Low-level: Give a byte array, and get a signature. Purpose: This just implements the signing
+//!   of byte arrays. Digital signature schemes generally just care about a set of input bytes to
+//!   sign; and this operation implements that per-supported digital signature scheme. To add
+//!   support for a new scheme, only this operation needs to be implemented for the new signing key
+//!   type.
+//!
+//! Further, there are two kinds of signing operations supported here:
+//! - Sign: Create a signed object that contains the payload. Purpose: If only one signature is
+//!   needed for an object then it is simpler to keep the signature and payload together in one
+//!   blob, so they cannot be separated.
+//! - Sign detached: Create a signature that does not contain the payload; but the serialized
+//!   payload is returned. Purpose: If multiple signatures are needed for one object, then sign
+//!   detached can be used.
+
+use ciborium::value::Integer;
+use coset::iana::CoapContentFormat;
+use ed25519_dalek::Signer;
+use serde::{de::DeserializeOwned, Serialize};
+
+use super::SigningNamespace;
+use crate::{
+    cose::SIGNING_NAMESPACE, error::SignatureError, CryptoError, RawSigningKey, RawVerifyingKey,
+    Signature, SignedObject, SigningKey, VerifyingKey,
+};
+
+impl SigningKey {
+    /// Signs the given payload with the signing key, under a given [`SigningNamespace`].
+    /// This returns a [`Signature`] object, that does not contain the payload.
+    /// The payload must be stored separately, and needs to be provided when verifying the
+    /// signature.
+    ///
+    /// This should be used when multiple signers are required, or when signatures need to be
+    /// replaceable without re-uploading the object, or if the signed object should be parseable
+    /// by the server side, without the use of COSE on the server.
+    /// ```
+    /// use bitwarden_crypto::{SigningNamespace, SignatureAlgorithm, SigningKey};
+    /// use serde::{Serialize, Deserialize};
+    ///
+    /// #[derive(Serialize, Deserialize, Debug, PartialEq)]
+    /// struct TestMessage {
+    ///  field1: String,
+    /// }
+    ///
+    /// let signing_key = SigningKey::make(SignatureAlgorithm::Ed25519).unwrap();
+    /// let message = TestMessage {
+    ///  field1: "Test message".to_string(),
+    /// };
+    /// let namespace = SigningNamespace::ExampleNamespace;
+    /// let (signature, serialized_message) = signing_key.sign_detached(&message, &namespace).unwrap();
+    /// // Verification
+    /// let verifying_key = signing_key.to_verifying_key();
+    /// assert!(verifying_key.verify_signature(&serialized_message.as_ref(), &namespace, &signature));
+    /// ```
+    #[allow(unused)]
+    pub fn sign_detached<Message: Serialize>(
```

Comment:
Note: These must be pub because otherwise we cannot use doctests as usage examples. It is important to me to have runnable working examples, but i would prefer to keep these pub(crate). Not sure what to do here...